dev/DynamicDependency/WinRTPackage.cpp

```diff
@@ -0,0 +1,231 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+#include "pch.h"
+
+#include "WinRTPackage.h"
+
+MddCore::WinRT::ThreadingModel MddCore::WinRTPackage::GetThreadingModel(
+    const std::wstring& activatableClassId)
+{
+    // Find the activation factory
+    if (!m_inprocModules.empty())
+    {
+        for (size_t index = 0; index < m_inprocModules.size(); ++index)
+        {
+            auto& inprocModule{ m_inprocModules[index] };
+
+            auto threadingModel{ inprocModule.Find(activatableClassId) };
+            if (threadingModel != MddCore::WinRT::ThreadingModel::Unknown)
+            {
+                return threadingModel;
+            }
+        }
+    }
+
+    // Not found
+    return MddCore::WinRT::ThreadingModel::Unknown;
+}
+
+void* MddCore::WinRTPackage::GetActivationFactory(
+    HSTRING className,
+    const std::wstring& activatableClassId,
+    REFIID iid)
+{
+    // Find the activation factory
+    if (!m_inprocModules.empty())
+    {
+        for (size_t index = 0; index < m_inprocModules.size(); ++index)
+        {
+            auto& inprocModule{ m_inprocModules[index] };
+
+            auto threadingModel{ inprocModule.Find(activatableClassId) };
+            if (threadingModel != MddCore::WinRT::ThreadingModel::Unknown)
+            {
+                //TODO change to return shared_ptr<inprocModule> rather than void*factory
```

Comment:
log issue

Reply:
There's an issue around removing things from the list while in use. Changing this to shared_ptr<> will fix _this_ but there's some larger concerns posed by URFW's implementation.

The problems can be summed up as URFW has no removal logic and is entirely lock-free.

The latter is a problem for our use. I solved it via the WinRTModuleManager using s_lock AND modified to be separate PackageGraphManager. Originally I added this data to PackageGraph objects and it worked great -- until reentrancy happened. Quite quickly, thank you unit tests :-) Call `GetActivationFactory(PackageManager)` and pkgmgr's implementation calls GetActivationFactory. Reentrancy and std::mutex isn't recursive so...boom. Callstack made the problem clear once I understood what I was looking at. Moving the WinRT support into a parallel stack under WinRTModuleManager solved that.

There's a problem around removal operations. Changing this to shared_ptr<> may not be enough. How deep the rabbit hole goes remains to be seen. URFW's implementation has a few comments relating to this (once you immersive your brain in it). I'll be filing an Issue shortly to address this in a future PR, but the shape of the solution is TBD pending a chat with @kythant  and @BenJKuhn

Reply:
Got it.  Please log an issue for this, thanks.